### PR TITLE
[PropTypes] Add warnings if PropTypes return functions

### PR DESCRIPTION
**Pull request by ariabuckles**
Summary:
Right now, if a component specifies a propType as, for example, `myProp: React.PropTypes.shape`, without an actual shape parameter, any prop type will be accepted, because `React.PropTypes.shape` returns a function (the actual validator), not an Error, currently indicating that propType checking passed.

This can create an unfortunate situation where a component looks like it has fully specified `propTypes`, but in fact does not.

This commit addresses this by warning if a propType checker returns anything non-falsy that is not an Error (currently all the library PropTypes return null or an Error).

Test Plan:
Added a unit test; ran `jest` in the root repo directory.
Also ran `grunt lint` and `grunt test`

### CODE REVIEW

File: src/classic/element/__tests__/ReactElementValidator-test.js [hunk 1]
```diff
@@ -362,6 +362,32 @@ describe('ReactElementValidator', function() {
...
+      }
+    });
+
+    ReactTestUtils.renderIntoDocument(
+      React.createElement(Component, {prop: {value: 'hi'}})
+    );
+
+    expect(console.warn.calls.length).toBe(1);
+    expect(console.warn.calls[0].args[0]).toBe(
+      'Warning: Component: the type of prop `prop` is invalid; the type ' +
```
Comment (spicyj) [inline src/classic/element/__tests__/ReactElementValidator-test.js]: nit: call the prop something other than `prop` so we can distinguish these two arguments?

File: src/classic/element/ReactElementValidator.js [hunk 2]
```diff
@@ -247,6 +247,18 @@ function checkPropTypes(componentName, propTypes, props, location) {
       } catch (ex) {
         error = ex;
       }
+      warning(
+        !error || error instanceof Error,
+        '%s: the type of %s `%s` is invalid; the type checker function ' +
```
Comment (spicyj) [inline src/classic/element/ReactElementValidator.js]: Maybe "type checker … is invalid" or "type specification … is invalid"? I could see misinterpreting this to mean that an invalid value was passed.

**Comment by spicyj** [comment]
lgtm otherwise, thank you!

File: src/classic/element/__tests__/ReactElementValidator-test.js (hunk 1 above)
Comment (ariabuckles) [inline src/classic/element/__tests__/ReactElementValidator-test.js]: good idea :)

File: src/classic/element/ReactElementValidator.js (hunk 2 above)
Comment (ariabuckles) [inline src/classic/element/ReactElementValidator.js]: yeah, i'll reword this to try to be less confusing

**Comment by ariabuckles** [comment]
I'll make those changes; thanks!

**Comment by spicyj** [comment]
Thank you!

**Comment by ariabuckles** [comment]
yay thanks!
